### DISCUSSION TITLE
CommandBar Question

### DISCUSSION THREAD

**Discussion by mschoneman** [BlazorFluentUI/BlazorFluentUI#377] [Q&A] [answered]
So far, I'm still learning Blazor so apologies if this is a dumb question.  I'm trying to use a `CommandBar` and I'd like to use databinding on some of the properties of the various `CommandBarItem`s.  For example, I want to disable the command while it is executing.  I've managed to make it all work but involves have a variable for each `CommandBarItem` and specifically setting the `Disabled` property to `true` and calling `StateHasChanged()`.  When I have a regular `button` I can just bind in the markup and it works.  

Compare `CommandBar` version:

```
<CommandBar Items="@CommandBarItems">
</CommandBar>

@code {
    private List<CommandBarItem> CommandBarItems;
    private CommandBarItem RefreshCommandBarItem;

    protected override Task OnInitializedAsync()
    {
        RefreshCommandBarItem = new CommandBarItem
        {
            Text = "Refresh",
            IconName = "Refresh",
            OnClick = async (args) =>
            {
                await Refresh();
            },
            Disabled = false,
        };


        CommandBarItems = new List<CommandBarItem>() { RefreshCommandBarItem };
        
        return Task.CompletedTask;
    }

    private async Task Refresh()
    {
        RefreshCommandBarItem.Disabled = true;
        StateHasChanged();
        try
        {
            //  Omitted
        }
        finally
        {
            RefreshCommandBarItem.Disabled = false;
            StateHasChanged();
        }
    }
}
```
vs regular button:
```
<button class="btn btn-primary" disabled="@IsRefreshing" @onclick="Refresh">Refresh</button>

@code {
    private bool IsRefreshing = false;

    private async Task Refresh()
    {
        IsRefreshing = true;
        try
        {
            //  Omitted
        }
        finally
        {
            IsRefreshing = false;
        }
    }
}
```

Am I missing something?  Thanks for any help!

**Comment by limefrogyank** [comment, marked as answer]
Hopefully, this is not too late.  If I am thinking about this correctly, because we are using a template to generate your buttons in the commandbar, you can't databind to them the way you would a regular button that you'd create in razor code.  You could define your own template in `ItemTemplate` but I'm not sure it would be easier this way.

Since this is C#, we do have a `Command` property that automatically disables the button if you're using `ICommand` correctly.  There's a simple example of how to use `Command` (and CommandParameter) in the Button page at the bottom.   But if you implement ICommand yourself (https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.icommand?view=net-5.0), you can define when the button can execute and when it cannot.  My favorite library for using these is ReactiveUI but it's not for everyone.  

If you don't want to implement your own, we include a class called `RelayCommand` that takes two parameters.  The first is an action that executes when you run the command (i.e. hit the button).  The second parameter is a callback that takes the current CommandParameter and returns true if it's allowed to execute or false if not (button disabled).

**Reply by mschoneman** [reply]
Perfect!  Thanks for taking the time to explain.
